lib/index.js

```diff
@@ -44,53 +44,113 @@ function PrismPlugin(opts) {
-        const string = text.text;
         const grammarName = opts.getSyntax(block);
         const grammar = Prism.languages[grammarName];
+        let tokens;
 
         if (!grammar) {
             return text.characters;
         }
 
-        const tokens = Prism.tokenize(string, grammar);
-        let offset = 0;
+        if (opts.blockPerLine && block === previousBlock) {
+            tokens = previousTokens;
+        }
+
+        else if (opts.blockPerLine) {
+            const string = block.nodes.toArray().map(n => n.text).join('\n');
+            tokens = Prism.tokenize(string, grammar);
+            previousBlock = block;
+            previousTokens = tokens;
+        }
+
+        else {
+            const string = text.text;
+            tokens = Prism.tokenize(string, grammar);
+        }
+
+        let o = 0;
+        let start = opts.blockPerLine ? block.getOffset(text.key) : 0;
+        let end = start + text.text.length;
+
+        function addMark(s, e, className) {
+            className = 'prism-token token ' + className;
 
-        function addMark(start, end, className) {
-            for (let i = start; i < end; i++) {
+            for (let i = s; i < e; i++) {
                 let char = characters.get(i);
                 let { marks } = char;
                 marks = marks.add(Slate.Mark.create({ type: MARK_TYPE, data: { className } }));
-                char = char.merge({ marks });
+                char = char.set('marks', marks);
                 characters = characters.set(i, char);
             }
         }
 
+        function getString(token) {
+            if (typeof token === 'string') {
+                return token;
+            }
+
+            if (typeof token.content === 'string') {
+                return token.content;
+            }
+
+            return token.content.map(getString).join('');
+        }
+
         function processToken(token, accu) {
             accu = accu || '';
 
+            if (typeof token !== 'string' && typeof token.content !== 'string') {
+                accu = accu + ' ' + token.type + ' ' + (token.alias || '');
+
+                for (var i =0; i < token.content.length; i++) {
+                    processToken(token.content[i], accu);
+                }
+
+                return
+            }
+
+            const string = getString(token);
+            let length = string.length;
+
+            if (opts.blockPerLine) {
+                const newlines = (string.match(/\n/mg) || []).length;
+                length -= newlines;
+            }
+
+            const offset = o;
+            const nextOffset = offset + length;
+            o = nextOffset;
+
+            if (offset < start && nextOffset <= start) {
+                return;
+            }
+
+            if (offset >= end) {
+                return;
+            }
+
+            const gap = Math.max(start - offset, 0)
+            let s = offset - start + gap;
+            let e = nextOffset - start + gap;
```

Comment:
I think there's a bug here if `nextOffset > end`

Testing with this text:

```html
<!-- Some 
HTML -->
```

We have a call `addMark(0, 18, 'comment')` on the first text line, that fails because the comment mark extends beyond the end of the text.